locale.c

```diff
@@ -6513,305 +6525,310 @@ S_my_langinfo_i(pTHX_
-         *      UTF-8, assume the locale is UTF-8.
-         *  4)  Otherwise, assume the locale isn't UTF-8
-         *
-         * To save cycles, if the locale name indicates it is a UTF-8 locale,
-         * we stop looking at the first instance with legal non-ASCII UTF-8.
-         * It is very unlikely this combination is coincidental. */
+    if (utf8ness) {
+        *utf8ness = is_utf8;
+    }
 
-        utf8ness_t strings_utf8ness = UTF8NESS_UNKNOWN;
-        char * scratch_buf = NULL;
-        Size_t scratch_buf_size = 0;
+    return retval;
 
-        /* List of strings to look at */
-        const int trials[] = {
+#  endif    /* All the implementations of my_langinfo() */
 
-#          if defined(USE_LOCALE_MONETARY) && defined(HAS_LOCALECONV)
+/*--------------------------------------------------------------------------*/
 
-            /* The first string tried is the locale currency name.  Often that
-             * will be in the native script.
-             *
-             * But this is usable only if localeconv() is available, as that's
-             * the way we find out the currency symbol. */
+}   /* my_langinfo() */
 
-            CRNCYSTR,
+#  ifndef HAS_DEFINITIVE_UTF8NESS_DETERMINATION
 
-#          endif
-#          ifdef USE_LOCALE_TIME
+STATIC const char *
+S_override_codeset_if_utf8_found(pTHX_ const char * codeset,
+                                       const char * locale)
+{
+#    define NAME_INDICATES_UTF8       0x1
+#    define MB_CUR_MAX_SUGGESTS_UTF8  0x2
 
-        /* We can also try various strings associated with LC_TIME, like the
-         * names of months or days of the week */
+    /* Override 'codeset' with UTF-8 if this routine guesses that it should be.
+     * Conversely (but rarely), "UTF-8" in the locale name might be wrong.  We
+     * return "" as the code set name if we find that to be the case.  */
 
-            DAY_1, DAY_2, DAY_3, DAY_4, DAY_5, DAY_6, DAY_7,
-            MON_1, MON_2, MON_3, MON_4, MON_5, MON_6, MON_7, MON_8,
-                                        MON_9, MON_10, MON_11, MON_12,
-            ALT_DIGITS, AM_STR, PM_STR,
-            ABDAY_1, ABDAY_2, ABDAY_3, ABDAY_4, ABDAY_5, ABDAY_6, ABDAY_7,
-            ABMON_1, ABMON_2, ABMON_3, ABMON_4, ABMON_5, ABMON_6,
-            ABMON_7, ABMON_8, ABMON_9, ABMON_10, ABMON_11, ABMON_12
+    unsigned int lean_towards_being_utf8 = 0;
+    if (is_codeset_name_UTF8(codeset)) {
+        lean_towards_being_utf8 |= NAME_INDICATES_UTF8;
+    }
 
-#          endif
-        };
+    /* For this portion of the file to compile, some C99 functions aren't
+     * available to us, even though we now require C99.  So, something must be
+     * wrong with them.  The code here should be good enough to work around
+     * this issue, but should the need arise, comments in S_is_locale_utf8()
+     * list some alternative C99 functions that could be tried.
+     *
+     * But MB_CUR_MAX is a C99 construct that helps a lot, is simple for a
```

Comment:
`MB_CUR_MAX` is C89

Reply:
And so it is, even though it isn't in K&R v2, based on draft C89.  And the Linux man pages say C99.  Changed via a93c9128eb1306fc15c7b25040919300ce889d03.

I see now that `mbtowc()` is also C89.  Again the man page says C99 and it is missing from K&R.